File: studyWorkbook/시뮬레이션/15898.py
import sys, copy
from itertools import permutations
from collections import defaultdict

input = sys.stdin.readline

def deepcopy(matrix):
    ret = []
    for z in range(2):
        ret.append([])
        for y in range(5):
            ret[-1].append([])
            for x in range(5):
                ret[-1][-1].append(matrix[z][y][x])

    return ret



def getQuality(kiln):
    qualityDic = defaultdict(int)

    for y in range(5):
        for x in range(5):
            qualityDic[kiln[1][y][x]] += kiln[0][y][x]
    return qualityDic['R'] * 7 + qualityDic['B'] * 5 + qualityDic['G'] * 3 + qualityDic['Y'] * 2


def recur(depth, kiln, order):
    global answer
    if depth >= 3:
        answer = max(answer, getQuality(kiln))
        return

    for y in range(5):
        for x in range(5):
            if not isValidRange(x, y):
                continue
            for i in range(4):
                materialInfo = materialDict[order[depth]][i]
                newKiln = deepcopy(kiln)
                putMaterials(newKiln, materialInfo['materials'], materialInfo['materialsInfo'], x, y)
                recur(depth + 1, newKiln, order)


def isValidRange(x, y):
    return 0 <= x <= 1 and 0 <= y <= 1

def rotate(matrix):
    rotatedMatrix = []
    for x in range(4):
        rotatedMatrix.append([])
        for y in range(3, -1, -1):
            rotatedMatrix[-1].append(matrix[y][x])

    return rotatedMatrix

def putMaterials(kiln, materials, materialsInfo, sx, sy):
    for y in range(sy, sy + 4):
        for x in range(sx, sx + 4):
            newQuality = kiln[0][y][x] + materials[y - sy][x - sx]
            if newQuality < 0:
                kiln[0][y][x] = 0
            elif newQuality > 9:
                kiln[0][y][x] = 9
            else:
                kiln[0][y][x] = newQuality

            if materialsInfo[y - sy][x - sx] == "W":
                continue
            else:
                kiln[1][y][x] = materialsInfo[y - sy][x - sx]


n = int(input())
materialDict = {}

for i in range(n):
    materials = []
    materialsInfo = []

    for _ in range(4):
        materials.append(list(map(int, input().split())))

    for _ in range(4):
         materialsInfo.append(list(input().split()))


    materialDict[i] = [{
        "materials": materials,
        "materialsInfo": materialsInfo
    }]

    for _ in range(3):
        materials = rotate(materials)
        materialsInfo = rotate(materialsInfo)
        materialDict[i].append({
            "materials": materials,
            "materialsInfo": materialsInfo
        })

case = [i for i in range(n)]
answer = 0
for comb in permutations(case, 3):
    kiln = [[[0 for _ in range(5)] for _ in range(5)] for _ in range(2)]
    recur(0, kiln, comb)

print(answer)

File: studyWorkbook/시뮬레이션/15898.js
const fs = require("fs");
const inputPath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const stdin = fs
  .readFileSync(inputPath)
  .toString()
  .split("\n")
  .map((line) => line.trim());

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const permutations = (array, maxDepth) => {
  const visited = Array(array.length).fill(0);
  const perm = [];

  const recur = (depth, path) => {
    if (depth >= maxDepth) {
      perm.push(path);
      return;
    }

    for (let i = 0; i < array.length; i++) {
      if (visited[i] == 0) {
        visited[i] = 1;
        recur(depth + 1, [...path, i]);
        visited[i] = 0;
      }
    }
  };

  recur(0, []);
  return perm;
};

const deepcopy = (matrix) => JSON.parse(JSON.stringify(matrix));

const rotate = (matrix) => {
  const rotatedMatrix = [];

  for (let x = 0; x < 4; x++) {
    rotatedMatrix.push([]);
    for (let y = 3; y >= 0; y--) {
      rotatedMatrix.at(-1).push(matrix[y][x]);
    }
  }

  return rotatedMatrix;
};

const recur = (depth, kiln, order) => {
  if (depth >= 3) {
    return;
  }

  for (let y = 0; y < 5; y++) {
    for (let x = 0; x < 5; x++) {
      if (!(x >= 0 && x <= 1 && y >= 0 && y <= 1)) {
        continue;
      }
      for (let i = 0; i < 4; i++) {
        const materialInfo = materialDict[order[depth]][i];
        const newKiln = deepcopy(kiln);
      }
    }
  }
};

const putMaterials = (kiln, materials, materialsInfo, sx, sy) => {
  for (let y = sy; y < sy + 4; y++) {
    for (let x = sx; x < sx + 4; x++) {
      const newQuality = kiln[0][y][x] + materials[y - sy][x - sx];
      if (newQuality < 0) {
        kiln[0][y][x] = 0;
      } else if (newQuality > 9) {
        kiln[0][y][x] = 9;
      } else {
        kiln[0][y][x] = newQuality;
      }

      if (materialsInfo[y - sy][x - sx] === "W") {
        continue;
      }
      kiln[1][y][x] = materialsInfo[y - sy][x - sx];
    }
  }
};

const solution = () => {
  const n = Number(input());
  const materialDict = new Map();

  for (let i = 0; i < n; i++) {
    const materials = [];
    const materialsInfo = [];

    for (let j = 0; j < 4; j++) {
      materials.push(input().split(" ").map(Number));
    }

    for (let j = 0; j < 4; j++) {
      materialsInfo.push(input().split(" ").map(Number));
    }

    materialDict[i] = [
      {
        materials,
        materialsInfo,
      },
    ];
    let rotatedMaterials = deepcopy(materials);
    let rotatedMaterialsInfo = deepcopy(materialsInfo);

    for (let j = 0; j < 3; j++) {
      rotatedMaterials = rotate(rotatedMaterials);
      rotatedMaterialsInfo = rotate(rotatedMaterialsInfo);
      materialDict[i].push({
        materials: rotatedMaterials,
        materialsInfo: rotatedMaterialsInfo,
      });
    }
  }

  const materialCase = Array.from({ length: n }, (_, index) => index);
  let answer = 0;
  permutations(materialCase, 3).forEach((perm) => {
    const kiln = Array.from({ length: 2 }, () =>
      Array.from({ length: 5 }, () => Array(5).fill(0))
    );
    // console.log(kiln);
    recur(0, kiln, perm);
  });
};

solution();
